lib/srv/db/mysql/mariadb_activate_user.sql

```diff
@@ -0,0 +1,91 @@
+    SET @teleport_user = JSON_VALUE(details, "$.attributes.user");
+    SET @all_in_one_role := CONCAT("tp-role-", username);
+
+    -- If the user already exists and was provisioned by Teleport, reactivate
+    -- it, otherwise provision a new one.
+    SELECT COUNT(USER) INTO is_auto_user FROM mysql.roles_mapping WHERE Role = 'teleport-auto-user' AND USER = username AND Admin_option = 'N';
+    IF is_auto_user = 1 THEN
+        SELECT COUNT(USER) INTO is_same_user FROM user_attributes WHERE USER = username AND JSON_VALUE(Attributes, "$.user") = @teleport_user;
+        IF is_same_user = 0 THEN
+            SIGNAL SQLSTATE 'TP001' SET MESSAGE_TEXT = 'Teleport username does not match user attributes';
+        END IF;
+
+        SELECT COUNT(USER) INTO is_active FROM information_schema.processlist WHERE USER = username;
+        -- If the user has active connections, make sure the provided roles
+        -- match what the user currently has.
+        IF is_active = 1 THEN
+            SELECT JSON_ARRAYAGG(Role) INTO cur_roles FROM mysql.roles_mapping WHERE USER = @all_in_one_role AND Admin_option = 'N';
+            SELECT JSON_EQUALS(@roles, cur_roles) INTO are_roles_same;
+            IF are_roles_same THEN
+                LEAVE proc_label;
+            ELSE
+                SIGNAL SQLSTATE 'TP002' SET MESSAGE_TEXT = 'user has active connections and roles have changed';
+            END IF;
+        END IF;
+
+        -- Ensure the user is unlocked. User is locked at deactivation.
+        SET @sql := CONCAT_WS(' ', 'ALTER USER', QUOTE(username), 'ACCOUNT UNLOCK');
+        PREPARE stmt FROM @sql;
+        EXECUTE stmt;
+        DEALLOCATE PREPARE stmt;
+    ELSE
+        SET @sql := CONCAT_WS(' ', 'CREATE USER', QUOTE(username), JSON_VALUE(details, "$.auth_options"));
+        PREPARE stmt FROM @sql;
+        EXECUTE stmt;
+        DEALLOCATE PREPARE stmt;
+
+        SET @sql := CONCAT_WS(' ', 'GRANT', QUOTE('teleport-auto-user'), 'TO', QUOTE(username));
+        PREPARE stmt FROM @sql;
+        EXECUTE stmt;
+        DEALLOCATE PREPARE stmt;
+
+        -- Set up the all-in-one role.
+        SET @sql := CONCAT_WS(' ', 'CREATE OR REPLACE ROLE', QUOTE(@all_in_one_role), 'WITH ADMIN CURRENT_USER');
+        PREPARE stmt FROM @sql;
+        EXECUTE stmt;
+        DEALLOCATE PREPARE stmt;
+
+        -- Set up user attributes.
+        CREATE TABLE IF NOT EXISTS user_attributes(User char(128), Attributes JSON, Primary Key (User));
+        CREATE OR REPLACE VIEW V_user_attributes AS SELECT * FROM user_attributes WHERE SUBSTRING_INDEX(SESSION_USER(),'@',1) = User ;
+        SET @attributes := JSON_EXTRACT(details,"$.attributes");
+        INSERT INTO user_attributes VALUES(username, @attributes) ON DUPLICATE KEY UPDATE Attributes=@attributes;
+        SET @sql := CONCAT_WS(' ', 'GRANT SELECT ON V_user_attributes TO', QUOTE(username));
+        PREPARE stmt FROM @sql;
+        EXECUTE stmt;
+        DEALLOCATE PREPARE stmt;
```

Comment:
As far as I can see, `V_user_attributes` is not actually used anywhere. We also grant `SELECT` on the view to the user, but I don't see why this is needed.

Overall, it is a bit strange to see table and view creation in the middle of all this; perhaps we should move this all away to the initialization part?

Reply:
correct. i was thinking providing a way for users to check their original identity name in their auto-provisioned sessions. not really necessary. will remove